Sources/Confidence/EventSenderEngine.swift

```diff
@@ -0,0 +1,97 @@
+import Combine
+import Foundation
+
+protocol EventsUploader {
+    func upload(request: EventBatchRequest) -> Bool
+}
+
+struct Event: Encodable, Equatable {
+    let name: String
+    let payload: [String: ConfidenceValue]
+}
+
+protocol FlushPolicy {
+    func reset()
+    func hit(event: Event)
+    func shouldFlush() -> Bool
+}
+
+protocol Clock {
+    func now() -> Date
+}
+
+protocol EventSenderEngine {
+    func send(name: String, message: [String: ConfidenceValue])
+    func shutdown()
+}
+
+final class EventSenderEngineImpl: EventSenderEngine {
+    private static let sendSignalName: String = "FLUSH"
+    private let storage: any EventStorage
+    private let writeReqChannel = PassthroughSubject<Event, Never>()
+    private let uploadReqChannel = PassthroughSubject<String, Never>()
+    private var cancellables = Set<AnyCancellable>()
+    private let flushPolicies: [FlushPolicy]
+    private let uploader: EventsUploader
+    private let clientSecret: String
+    private let clock: Clock
+
+    init(
+        clientSecret: String,
+        uploader: EventsUploader,
+        clock: Clock,
+        storage: EventStorage,
+        flushPolicies: [FlushPolicy]
+    ) {
+        self.clock = clock
+        self.uploader = uploader
+        self.clientSecret = clientSecret
+        self.storage = storage
+        self.flushPolicies = flushPolicies
+
+        writeReqChannel.sink(receiveValue: { [weak self] event in
+            guard let self = self else { return }
+            do {
+                try self.storage.writeEvent(event: event)
+            } catch {
+
+            }
+
+            self.flushPolicies.forEach({ policy in policy.hit(event: event) })
+            let shouldFlush = self.flushPolicies.contains(where: { policy in policy.shouldFlush() })
+
+            if shouldFlush {
+                self.uploadReqChannel.send(EventSenderEngineImpl.sendSignalName)
+                self.flushPolicies.forEach({ policy in policy.reset() })
+            }
+
+        }).store(in: &cancellables)
+
+        uploadReqChannel.sink(receiveValue: { [weak self] _ in
+            do {
+                guard let self = self else { return }
+                try self.storage.startNewBatch()
+                let ids = storage.batchReadyIds()
+                for id in ids {
+                    let events = try self.storage.eventsFrom(id: id)
+                    let batchRequest = EventBatchRequest(
+                        clientSecret: clientSecret, sendTime: clock.now(), events: events)
```

Comment:
would be nice to not have to append the `clientSecret` as part of every `EventBatchRequest`. Maybe have the uploader created with that instead.

Reply:
good thing about client secret being in the batchRequest object is it all will be encoded as part of `EventBatchRequest ` object. no more additional type is needed. so basically `EventBatchRequest` represents what backend accepts. we can create a network type that has this client secret and introduce another type without client secret that engine works with.

Reply:
ah, ok, yes that might be nice.. but I'm not sure about the engine creating it?

Reply:
yeah if we decide to have that additional type, that should be created in the uploader (and probably only visible to the network layer which is uploader)

Reply:
I also think it might be more safe to set the `sendTime` as part of the uploading if we put some retry mechanism between this and the uploader (or inside the uploader).

Reply:
i think it makes sense for engine to just send the list of events then and uploader attaches the network required types like send time client secret etc.